### DISCUSSION TITLE
How to fetch all followings of a user?

### DISCUSSION THREAD

**Discussion by emreond** [sbertix/Swiftagram#179] [Q&A] [answered]
Hello,

I know I asked much but I hope this will be my last :). I am trying to fetch all followings of a user (not the logged one) but It only returns 600. The user I want to fetch followings has 950 following. I am using `Friendship.following` endpoint for that. Am I doing something wrong? 

```
Endpoint.Friendship.following(identifier)
            .unlocking(with: secret)
            .task(maxLength: .max, by: .default, onComplete: { _ in
```

**Comment by sbertix** [comment, marked as answer]
Hey @emreond, 

It's either because Instagram rate limited your account (as you're making too many requests), or simply because you're expecting a different number (based on the counter you can see) when 600 it's actually the correct amount. 
Actual counters are not immediately updated every time an account is deactivated or suspended, and of course followers/following will not display users who blocked you, etc. This means that for an account with a counter of ~700 followers, you can expect ~600 "real" ones. 

If that's not the case, and for instance you're trying to paginate some A-list celebrity account, than rate limiting is at place: the Private API are used for displaying content driven by user interaction. As it is unlikely for a user to scroll even just 10 pages of a followers list, and it's extremely valuable data, followers/following endpoints are among the most limited ones.

**Reply by emreond** [reply]
Yea you are right, I guess. It is misleading. For example, with 750 following count user returns 1050 item when fetched with that endpoint :))  Also I found that some datas return twice, I was getting all followings with below code but I found that there are some duplicates when I just append all returned values. Do you use `Set` or something for deal with this problem?

```
        var followings: [Swiftagram.User] = []
        Endpoint.Friendship.following(identifier)
            .unlocking(with: secret)
            .task(maxLength: .max, by: .default, onComplete: { _ in
                completion(followings)
            }) { (result) in
                if let followingUsers = try? result.get().users {
                    followings.append(contentsOf: followingUsers)
                } 
            }
```

**Reply by emreond** [reply]
I tried with another account which is not used in any third-party apps but following api always returns different values. For example It returns a user with nick "A" in one request and on other It doesn't returned it. Actually I was trying to get the new followings via storing old ones in realm and calculation diff. May I just get first page with sort order like in Instagram App for our own followings?

**Reply by sbertix** [reply]
> I tried with another account which is not used in any third-party apps but following api always returns different values. For example It returns a user with nick "A" in one request and on other It doesn't returned it. Actually I was trying to get the new followings via storing old ones in realm and calculation diff. May I just get first page with sort order like in Instagram App for our own followings?

Sorting is not unique for followers/following. Different accounts get different sorting and you should not expect them to match, not even when using the same rank token. 
You're responsible for sorting them as you like once you've fetched them.

**Reply by emreond** [reply]
Then It is impossible what I want to achieve. (Fetch all followings and diff with old values) :(( Thats sad to hear.

**Reply by sbertix** [reply]
Not necessarily. Rank token should work, **as long as you're fetching followers/following from the same account**.

**Reply by emreond** [reply]
Is there any example about what rank token is?

**Reply by sbertix** [reply]
It's close to a seed for random number generators. 

Given…

```swift
let identifier: String = /* some valid Instagram user identifier */
```

… when you call `following(_:)`/`followed(by:)`, you can make sure pages are fetched without overlaps and demanding a pre-determined sorting, by specifying a rank token, which is just a constant string you can pick at random and save between fetches.

```swift
Endpoint.Friendship.following(identifier, rank: "somerandomconstantstring")
   […]
```

**Reply by emreond** [reply]
May I also learn what is `startingAt` value? If It is a dating value, which returns followings after a certain date. It also solved my problem like a rocket :D

**Reply by sbertix** [reply]
Unfortunately it's not. It's simply the `max_id` for the starting page. 
Meaning you can fetch one page, then get the pagination offset and use that as the argument of the method, and start from the second one instead of from scratch.

**Reply by emreond** [reply]
Aaa of course. By the way, rank is also didn't solve my problem. I tried to fetch all followings with below code (because It is not date sorted, I needed to fetch all followings) but when I call this method second time It doesn't return same response and this is because on first call It doesn't return all followings but just a part of them. 

**P.S** I can wait more than default request gap If It will solve my problem to fetch all followings or If there is some way to detect couldn't fetched followings, I can retry that page only.

`  Endpoint.Friendship.following(identifier, matching: nil, startingAt: nil, rank: "followingfetchforuser1")
            .unlocking(with: secret)
            .task(maxLength: .max, by: .default, onComplete: { _ in`

**Reply by sbertix** [reply]
Do you mean that if you write this code you get two different sortings?

```swift
Endpoint.Friendship.following(identifier, rank: "followingfetchforuser1")
   […]
// And then again with the same exact rank token and starting from the same page…
Endpoint.Friendship.following(identifier, rank: "followingfetchforuser1")
   […]
// … you get two different sortings?
```

That should not happen

**Reply by emreond** [reply]
Yes even with one page It returns different response. Let me try to give more detail example;

A user has 802 followings (returned from counter.following and seen on Instagram profile page) and when I fetch only one page of followings (It doesn't matter how many pages I fetch for the diff occur) with below code, It returns 100 value. Lets say in this 100 value, there is a user called "Ronaldo", when I call this function again **new** 100 value may or may not have "Ronaldo" and have "Emre" for example which is not in the first returned 100 value.

```
 Endpoint.Friendship.following(identifier, matching: nil, startingAt: nil, rank: "followingfetchforuser1")
            .unlocking(with: secret)
            .task(maxLength: 1, by: .default, onComplete: { _ in
```


**P.S:** When I try to fetch all followings with `.max` page value, It returns 618 value which is less than 802. I tried to fetch followings of accounts with  4, 750 and 802 following and I could only fetch all followings with 4 following account. In addition, I tried this process with two different account (To check If my own account is blocked or something for fetching).

Swiftagram Version: **4.2.10**

**Reply by emreond** [reply]
I will jump from the bridge. It was `.following` not `.followed`.
